sphinx/ext/autodoc/importer.py

```diff
@@ -84,22 +85,19 @@ def import_object(modname: str, objpath: list[str], objtype: str = '',
         while module is None:
             try:
                 orig_modules = frozenset(sys.modules)
+                import_module(modname, warningiserror=warningiserror)
+
+                # Try reloading modules with ``typing.TYPE_CHECKING == True``.
                 try:
-                    # try importing with ``typing.TYPE_CHECKING == True``
                     typing.TYPE_CHECKING = True
-                    module = import_module(modname, warningiserror=warningiserror)
-                except ImportError:
-                    # if that fails (e.g. circular import), retry with
-                    # ``typing.TYPE_CHECKING == False`` after reverting
-                    # changes made to ``sys.modules`` by the failed try
-                    for m in [m for m in sys.modules if m not in orig_modules]:
-                        sys.modules.pop(m)
-
-                    typing.TYPE_CHECKING = False
-                    module = import_module(modname, warningiserror=warningiserror)
+                    # Ignore failures; we've already successfully loaded these modules.
+                    with contextlib.suppress(Exception):
```

Comment:
Can we have a narrower exception type than Exception?

Reply:
I don't think so -- at least, not without sacrificing correctness. If I change this to just catch `ImportError`, then doing:
```rest
.. automodule:: pydantic
```
fails with::
```text
>       _PartialClsOrStaticMethod: TypeAlias = Union[classmethod[Any, Any, Any], staticmethod[Any, Any], partialmethod[Any]]
E       TypeError: type 'classmethod' is not subscriptable
```

Let me flip the question around on you: given that the module was already imported successfully with `typing.TYPE_CHECKING == False`, which sorts of exceptions encountered while reloading the module with `typing.TYPE_CHECKING == True` should cause `import_object` to fail instead of returning the already successfully imported module to our caller?

Reply:
Good point. I take it from your testing that e.g. if ``pydantic`` is successfully imported, and then fails on the reload, the initial imported module is still usable, coherent, etc?

Reply:
That's my expectation, and it seems to be the case for pydantic, at least up to the point where a basic `.. automodule:: pydantic` succeeds. I'm not sure how to test it more thoroughly, though. I do suspect that this is still going to be fragile in ways that we can't predict, the hope is just that it works for more cases than both what's in #11511 and what's in #11645